python/src/aac/lang/lsp/code_completion_provider.py

```diff
@@ -0,0 +1,89 @@
+class CodeCompletionProvider:
+    """Resolve various code compleetion triggers to specific code completion functions.
+
+    Attributes:
+        registered_completion_callbacks (dict[str, Callable]): The dict of trigger characters to corresponding code completion functions
+    """
+
+    registered_completion_callbacks: dict = attrib(default=Factory(dict), validator=validators.instance_of(dict))
+
+    def __attrs_post_init__(self):
+        """Post init hook for attrs classes."""
+        self.registered_completion_callbacks[SPACE_TRIGGER] = _handle_space_code_completion
+
+    def get_trigger_characters(self) -> list[str]:
+        """Return a list of the currently registered trigger characters."""
+        return list(self.registered_completion_callbacks.keys())
+
+    def handle_code_completion(self, ls: LanguageServer, params: CompletionParams):
+        """Resolve the trigger to the corresponding code completion function, then execute it."""
+        trigger_character = params.context.trigger_character
+
+        if not trigger_character:
+            source_document_content = ls.workspace.documents.get(params.text_document.uri).source.split(os.linesep)
+            trigger_character = source_document_content[params.position.line][params.position.character - 1]
+
+        callback_function = self.registered_completion_callbacks.get(trigger_character)
+        if callback_function:
+            return callback_function(ls, params)
+        else:
+            logging.debug(f"Failed to find corresponding code completion function for the registered trigger: {trigger_character} in registered callbacks: {self.registered_completion_callbacks}")
+
+
+def _handle_space_code_completion(ls: LanguageServer, params: CompletionParams):
+    source_document_content = _get_code_completion_parent_text_file(ls, params).split(os.linesep)
+    position_line = source_document_content[params.position.line]
+
+    if position_line.strip().startswith("type:"):
+        return _get_reference_completion_items(ls, params)
+
+
+def _get_reference_completion_items(ls: LanguageServer, params: CompletionParams):
+    active_context = get_active_context()
+    primitives_definition = active_context.get_primitives_definition()
+
+    primitive_references = {}
+    if primitives_definition:
+        primitive_references = {field: "Primitive type" for field in active_context.get_primitive_types()}
+
+    data_definition_references = _convert_definitions_to_name_description_dictionary(active_context.get_definitions_by_root_key("data"))
+
+    file_definitions = parse(_get_code_completion_parent_text_file(ls, params))
+    file_data_references = _convert_definitions_to_name_description_dictionary(get_definitions_by_root_key("data", file_definitions))
+
+    return CompletionList(
+        is_incomplete=False,
+        items=[
+            CompletionItem(
+                label=reference,
+                kind=CompletionItemKind.Reference,
+                documentation=""
```

Comment:
Is there a reason not to include the `documentation` string? Is it because the box is too small for the formatting we have in the string?

Reply:
At one point, but I forgot to add it back. I'll add it back in.

Reply:
I'm not seeing the description when it is populated, and not all definitions have a defined description.

Reply:
This image comes from a debug session when I'm on the `main` branch. (I'm assuming this is similar to what you're seeing):

![Screen Shot 2022-04-15 at 11 12 28](https://user-images.githubusercontent.com/2924081/163594392-880815f8-7c17-4e25-8504-2bdaf6557e46.png)

What I had to do to see the `documentation` string was hover over one of the options and click the arrow. Then you should be able to see the `documentation` strings for every option that has one without any further configuration:


![Screen Shot 2022-04-15 at 11 16 11](https://user-images.githubusercontent.com/2924081/163595040-62f96936-f2ff-4e6b-8c49-49e1379f9a86.png)

![Screen Shot 2022-04-15 at 11 14 56](https://user-images.githubusercontent.com/2924081/163594744-9e59356a-4624-44ce-9f39-cac6ba31b73f.png)

Reply:
Could you please add the `documentation` back?

Reply:
Done, but I still don't see the option to expand even though descriptions are being populated. See this log excerpt.
```
04/15/2022 19:03:37::protocol::protocol.py::L383::INFO::Sending data: {"jsonrpc": "2.0", "id": 1, "result": {"isIncomplete": false, "items": [{"label": "string", "kind": 18, "documentation": "Primitive type"}, {"label": "int", "kind": 18, "documentation": "Primitive type"}, {"label": "number", "kind": 18, "documentation": "Primitive type"}, {"label": "bool", "kind": 18, "documentation": "Primitive type"}, {"label": "date", "kind": 18, "documentation": "Primitive type"}, {"label": "file", "kind": 18, "documentation": "Primitive type"}, {"label": "map", "kind": 18, "documentation": "Primitive type"}, {"label": "root", "kind": 18, "documentation": ""}, {"label": "enum", "kind": 18, "documentation": ""}, {"label": "extension", "kind": 18, "documentation": ""}, {"label": "EnumExtension", "kind": 18, "documentation": ""}, {"label": "DataExtension", "kind": 18, "documentation": ""}, {"label": "Validation", "kind": 18, "documentation": ""}, {"label": "ValidationReference", "kind": 18, "documentation": ""}, {"label": "ValidatorInput", "kind": 18, "documentation": ""}, {"label": "ValidatorOutput", "kind": 18, "documentation": ""}, {"label": "Scenario", "kind": 18, "documentation": ""}, {"label": "Field", "kind": 18, "documentation": ""}, {"label": "data", "kind": 18, "documentation": ""}, {"label": "Behavior", "kind": 18, "documentation": ""}, {"label": "model", "kind": 18, "documentation": ""}, {"label": "usecase", "kind": 18, "documentation": ""}, {"label": "Step", "kind": 18, "documentation": ""}, {"label": "Branch", "kind": 18, "documentation": ""}, {"label": "Specification", "kind": 18, "documentation": ""}, {"label": "SpecSection", "kind": 18, "documentation": ""}, {"label": "Requirement", "kind": 18, "documentation": ""}, {"label": "RequirementAttribute", "kind": 18, "documentation": ""}, {"label": "RequirementReference", "kind": 18, "documentation": ""}, {"label": "DataA", "kind": 18, "documentation": ""}, {"label": "DataB", "kind": 18, "documentation": ""}, {"label": "DataC", "kind": 18, "documentation": ""}, {"label": "DataD", "kind": 18, "documentation": ""}, {"label": "MessageMetadataData", "kind": 18, "documentation": ""}]}}
```

Reply:
Done